lib/Frontend/CompilerInvocation.cpp

```diff
@@ -118,94 +118,80 @@ class ArgsToFrontendInputsConverter {
+  SmallVector<std::unique_ptr<llvm::MemoryBuffer>, 4> BuffersToKeepAlive;
 
   llvm::SetVector<StringRef> Files;
+  std::set<StringRef> PrimaryFiles;
 
 public:
   ArgsToFrontendInputsConverter(DiagnosticEngine &Diags, const ArgList &Args,
                                 FrontendInputs &Inputs)
       : Diags(Diags), Args(Args), Inputs(Inputs),
-        FilelistPathArg(Args.getLastArg(options::OPT_filelist)) {}
+        FilelistPathArg(Args.getLastArg(options::OPT_filelist)),
+        PrimaryFilelistPathArg(Args.getLastArg(options::OPT_primary_filelist)) {
+  }
 
   bool convert() {
     if (enforceFilelistExclusion())
       return true;
-    bool hadError = getFilesFromCommandLine();
-    if (hadError)
+    if (FilelistPathArg ? readInputFilesFromFilelist()
+                        : readInputFilesFromCommandLine())
       return true;
-    hadError = getFilesFromFilelist();
-    if (hadError)
+    if (readPrimaryFiles())
       return true;
-
-    llvm::StringSet<> PrimaryFiles = getPrimaries();
-
-    for (auto file : Files) {
-      bool isPrimary = PrimaryFiles.count(file) > 0;
-      Inputs.addInput(InputFile(file, isPrimary));
-      if (isPrimary)
-        PrimaryFiles.erase(file);
-    }
-    for (auto &file : PrimaryFiles) {
-      // Catch "swiftc -frontend -c -filelist foo -primary-file
-      // some-file-not-in-foo".
-      assert(doesCommandLineIncludeFilelist() &&
-             "Missing primary with no filelist");
-      Diags.diagnose(SourceLoc(), diag::error_primary_file_not_found,
-                     file.getKey(), filelistPath());
-    }
-    return !PrimaryFiles.empty();
+    createInputFiles();
+    return checkForMissingPrimaryFiles();
   }
 
 private:
   bool enforceFilelistExclusion() {
-    if (Args.hasArg(options::OPT_INPUT) && doesCommandLineIncludeFilelist()) {
+    if (Args.hasArg(options::OPT_INPUT) && FilelistPathArg) {
       Diags.diagnose(SourceLoc(),
                      diag::error_cannot_have_input_files_with_file_list);
       return true;
     }
     return false;
   }
 
-  bool getFilesFromCommandLine() {
+  bool readInputFilesFromCommandLine() {
     bool hadDuplicates = false;
     for (const Arg *A :
          Args.filtered(options::OPT_INPUT, options::OPT_primary_file)) {
-      if (A->getOption().matches(options::OPT_primary_file) &&
-          mustPrimaryFilesOnCommandLineAlsoAppearInFileList())
-        continue;
       hadDuplicates = addFile(A->getValue()) || hadDuplicates;
     }
     return false; // FIXME: Don't bail out for duplicates, too many tests depend
                   // on it.
   }
 
-  bool doesCommandLineIncludeFilelist() { return FilelistPathArg; }
-  bool mustPrimaryFilesOnCommandLineAlsoAppearInFileList() {
-    return doesCommandLineIncludeFilelist();
+  bool readInputFilesFromFilelist() {
+    bool hadDuplicates = false;
+    if (forAllFilesInFilelist(FilelistPathArg, [&](StringRef file) -> void {
+          hadDuplicates = addFile(file) || hadDuplicates;
```

Comment:
Personal preference: can you move this outside the `if` and store the result in a temporary? There's a bit too much going on here.

Reply:
Good idea. Done.